lib/kubernetes-deploy/kubernetes_resource/custom_resource.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+require 'jsonpath'
+module KubernetesDeploy
+  class CustomResource < KubernetesResource
+    def initialize(namespace:, context:, definition:, logger:, statsd_tags: [], crd:)
+      super(namespace: namespace, context: context, definition: definition,
+        logger: logger, statsd_tags: statsd_tags)
+      @rollout_config = crd.rollout_config
+    end
+
+    def deploy_succeeded?
+      return super unless @rollout_config
+      return false unless observed_generation == current_generation
+
+      @rollout_config.deploy_succeeded?(@instance_data)
+    end
+
+    def deploy_failed?
+      return super unless @rollout_config
+      return false unless observed_generation == current_generation
```

Comment:
Custom resource controllers are not required to implement the observed generation field, though it is a best practice. We should only make this comparison if the observed generation field is present.

Reply:
I'm more inclined to make `observedGeneration` be a necessary field for those taking advantage of this feature. Without it, there's no way of knowing what generation of the resource the status is referring to. In the worst case, we'll deploy breaking changes to a previously healthy resource and, since the old status will indicate success, the deploy will pass instantaneously before the controller has a chance to update the status and indicate something has gone wrong.

A possible workaround is to measure against `lastTransitionTime`, but we can't really guarantee that field will be there, either. I'm open to suggestions but I don't see another way around it.

FWIW I intend to update all the cloudbuddies `UpdateStatus` functions to transparently populate the `observedGeneration` and `lastTransitionTime` fields (example [here](https://github.com/Shopify/cloudbuddies/pull/1522/files#diff-c06febb409f662c2c8f104b739e4888cR18))

Reply:
You're definitely right that there are race conditions without that field--we saw them in the past, when even some core controllers didn't have it (and before we knew how important it is)! I don't think we should use any time-based logic. How do we require observed generation to be implemented though? Assume a user gets into the situation where they are correctly set up in every respect except this field--what do they see? One option would be to time out (which I think is the current behaviour with this PR), but make sure our error message explains that the observed generation field was missing / didn't match.